Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,517 @@
+            throw BSONError.InvalidArgumentError(message: "Syntax Error: Missing digits in front of the exponent")
+        }
+        let decimalPart = String(data[decimalPartRange])
+        var digits = try Self.convertToDigitsArray(decimalPart)
+
+        var exponent = 0
+        let exponentPartRange = match.range(at: REGroups.exponentPart.rawValue)
+        if exponentPartRange.location != NSNotFound, let range = Range(exponentPartRange, in: data) {
+            exponent = exponentSign * (Int(data[range]) ?? 0)
+        }
+        if let pointIndex = decimalPart.firstIndex(of: ".") {
+            // move the exponent by the number of digits after the decimal point
+            // so we are looking at an "integer" significand, easier to reason about
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+        }
+
+        while exponent > Self.exponentMax && digits.count <= Self.maxSignificandDigits {
+            // Exponent is too large, try shifting zeros into the coefficient
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < Self.exponentMin && !digits.isEmpty {
+            // Exponent is too small, try taking zeros off the coefficient
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = Self.exponentMin
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error: Value too small")
+            }
+        }
+
+        guard (Self.exponentMin...Self.exponentMax).contains(exponent) else {
+            throw BSONError.InvalidArgumentError(message: "Rounding Error: Cannot round exponent \(exponent) further")
+        }
+
+        guard digits.count <= Self.maxSignificandDigits else {
+            throw BSONError.InvalidArgumentError(
+                message: "Overflow Error: Value cannot exceed \(Self.maxSignificandDigits) digits"
+            )
+        }
+
+        let significandLoDigits = [UInt8](digits.suffix(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+        let significandHiDigits = [UInt8](digits.dropLast(Self.maxSignificandDigits / 2)).decimalDigitsToUInt64()
+
+        // Multiply by one hundred quadrillion (note the seventeen zeroes)
+        // the product is the significandHiDigits "shifted" up by 17 decimal places
+        // we can then add the significandLoDigits to the product to ensure that we have a correctly formed significand
+        var significand = UInt128.multiply(significandHiDigits, by: Self.decimalShift17Zeroes)
+        significand.lo += significandLoDigits
+
+        if significand.lo < significandLoDigits {
```

Comment:
can you explain how/why we encounter this case, and why adding 1 to the significand handles it? since significand.lo is a UInt64 I don't see how we end up with a value here that is less than significandLoDigits after adding significandLoDigits.